Native Command Error Handling (#261)
PowerShell/PowerShell-RFC: 1-Draft/RFC00XX-Native-Command-Error-Handling.md

---
RFC: RFC00XX
Author: Jason Helmick
Status: Draft
Area: Core
Comments Due: 10/31/2020
---

# Native Command Error Handling

PowerShell scripts using native commands would benefit from being able to use error handling
features like those used by cmdlets.

## Motivation

In PowerShell by default, script processing continues when non-terminating errors occur. This is a
benefit when expecting non-terminating errors in normal execution such as non-responsive computers
from a list. This default behavior is controlled with the preference variable
`$ErrorActionPreference` default of `Continue`.

> In production, often customers prefer that script execution stops when a non-terminating error

Reviewer: ```suggestion
In production, often customers prefer that script execution stops when any error, including non-terminating errors, 
```

occurs. This is particularly true in CI where the preference is to fail fast. PowerShell currently
supports customers with this ability by setting the `$ErrorActionPreference` variable in the script
to `Stop`.

```Powershell
$ErrorActionPreference = 'Stop'
```

Native commands usually return an exit code to the calling application which will be zero for
success or non-zero for failure. However, native commands currently do not participate in the

> PowerShell error stream. Redirected `stderr` output is not interpreted the same as the PowerShell

Reviewer: stderr not being interpreted as an error by $ErrorActionPreference was a new experimental feature added to 7.1

error stream as many native commands use `stderr` as information/verbose stream and thus only the
exit code matters. Users working with native commands in their scripts will need to check the
execution status after each call using a helper function similar to below:

```Powershell

> if ($LASTEXITCODE -ne 0)

Reviewer 2: Worth mentioning the v7+ alternative based on the pipeline-chain operators:

```
ls /no/such/path || $(throw "Command failed. See above errors for details") # note the need for $(...)
```

{
    throw "Command failed. See above errors for details"
}
 ```

Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
stale.

> In POSIX shells, this need to terminate on command error is addressed by the `set -e` configuration,

Reviewer 2: I think it is implied by the current revision, but let me spell it out to see if my understanding is correct:

* _Unlike_ `bash`'s `set -e`, **PowerShell will _not_ (and should not) make the failure detection dependent on whether or not it is used in a _conditional_** (e.g., of an `if` statement).

   * Unlike in POSIX-like shells, where conditionals act on _exit codes_, PowerShell's conditionals act on _data_, so it make sense to treat failure detection separately and to thereby  avoid the rule complexity surrounding `set -e`.

* _Like_ `bash`'s `set -e`, **PowerShell _will_ bypass failure detection if a failure occurs in a non-final segment of a _pipeline chain_**, so as to allow overriding `$PSNativeCommandErrorAction` _ad hoc_.

Examples:

```powershell
$PSNativeCommandErrorAction = 'Stop'

# SHOULD fail, if `ls` reports a nonzero exit code.
if (-not ($name = /bin/ls nosuch.txt))  {
  $name = 'default.txt'
}

# Should NOT fail, because || is used to explicitly handle the failure:
/bin/ls nosuch.txt || 'default.txt'
```

> which causes the shell to exit when a command fails. In addition, to ensure that an error is

Reviewer: ```suggestion
which causes the shell to exit when a command fails due to non-zero exit code. In addition, to ensure that an error is
```

> returned if any command in a pipeline fails, POSIX shells address this need with `set -o pipefail`

Reviewer: I think it would be good to have an example script showing the difference in behavior between pipefail and w/o pipefail

objects that are added to the error stream and may terminate execution of the script without added
boilerplate.

The specification proposes similar functionality to the common POSIX shell configuration `set -eo pipefail`.

- `set -e` - terminates on command error
- `set -o pipefail` - returns an error if any command in the pipeline fails.

> [!NOTE] A common configuration command for POSIX shells `set -euo pipefail` includes the `set -u`
> configuration which returns an error if any variable has not been previously defined. This is
> equivalent to the existing PowerShell `Set-StrictMode` and is not addressed in this RFC.

The `$PSNativeCommandErrorAction` preference variable will implement a version of the
`$ErrorActionPreference` variable for native commands.

- The value will default to `Ignore` for compatibility with existing behavior.
- For non-zero exit codes and except for the value `Ignore`, an `ErrorRecord` will be added to
  `$Error` that wraps the exit code and the command executed that returned the exit code.
- Initially, only the existing values of `$ErrorActionPreference` will be supported.

> - The set of values may be extended later to include `MatchErrorActionPreference`, which should

Reviewer: Should move this to Considerations section

  apply the `$ErrorActionPreference` setting to native commands.
- A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference`
  values, where `MatchErrorActionPreference` is converted to the current value of
  `$ErrorActionPreference`

Valid values for `$PSNativeCommandErrorAction`

| Value           | Definition
----------------  | -------------------
| Break           | Enter the debugger when an error occurs or when an exception is raised.

> | Continue        | (Default) - Displays the error message and continues executing.

Reviewer: There's a mismatch here between Continue being default and line 83

| Ignore          | Suppresses the error message and continues to execute the command.
| Inquire         | Displays the error message and asks you whether you want to continue.

> | SilentlyContinue| No effect. The error message isn't displayed and execution continues without interruption.

Reviewer: Not really "no effect" as the error record is added to $Error

> | Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream

Reviewer 2: Typo: "stream" is doubled.

> | Suspend         | Automatically suspends a workflow job to allow for further investigation.

Reviewer: Suspend is not supported as it's only for workflows

The reported error record object will be the new type: `NativeCommandException` with the following details:

| Property        | Definition
----------------  | -------------------
| ExitCode:      |  The exit code of the failed command.
| ErrorID:       | `"Program {0} ended with non-zero exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.

> | ErrorCategory: | `ErrorCategory.NotSpecified`.

Reviewer 2: Perhaps it's worth creating a specific category for this, such as `NativeFailure` - not sure how that would affect backward compatibility (at least in PowerShell code it shouldn't).

| object:         | exit code
| Source:        | The full path to the application
| ProcessInfo    | details of failed command including path, exit code, and PID

> ## Alternative Approaches and Considerations

Reviewer: Add that additional configuration may be enabled to disable pipefail if there is a customer need for it

> One way of checking for a single native command and handling its exit

Reviewer: Include that on Windows, many native commands do not strictly follow the non-zero is error, so we may add a special list to handle those commands differently where known non-zero exit code is not treated as an error.

Reviewer 2: @SteveL-MSFT, I suggest we not do that, because it adds complexity while providing little benefit: the "rogue" CLIs fall into two categories:

* True rogue CLIs that simply ignore exit codes and _always_ return `0`
  * There's nothing we can generally do to compensate for that, short of - brittlely - trying to analyze stderr output, which I don't think is worth doing.

* CLIs such as `robocopy.exe` that repurpose specific nonzero exit codes to communicate additional information  about _success_ conditions (while also using (other) nonzero exit codes to communicate failure).

 * These exit codes are CLI-specific, and the user needs to know about them and handle them; while we _could_ hard-code exceptions for the nonzero exit codes that signal _success_, the problems are:
    * Do we really want to code such exceptions for _all_ utilities that ship with Windows (assuming there are others - I don't know)? 
    * Users may expect the same magic to apply to _all_ utilities, whether in-box or not.
    * Maintenance burden: What if utilities later introduce _additional_ nonzero exit codes that aren't failures?

Instead, I think we should make it as easy as possible to ignore a nonzero exit code _ad hoc_, which can already be done, though it's not the most obvious solution:

```powershell
# Call robocopy and ignore any failure that a nonzero exit code may trigger.
robocopy ... || @()

# ... now analyze $LASTEXITCODE
```

Reviewer 3: msiexec has two values that indicate success, but that a reboot is required (1641 and 3010), along with 0 also indicating success. That's the other one that I run into constantly (besides robocopy).

Discussed here: https://docs.microsoft.com/en-us/windows/win32/msi/error-codes

I'm sure there are more.

Reviewer 2: Thanks, @essentialexch, that's good to know.

However, instead of getting into the hard-coded exception business, I suggest _documenting_ such high-profile exceptions as part of  a _conceptual help topic dedicated to external-program calls_: see https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5152

---
RFC: RFC00XX
Author: Jason Helmick
Status: Draft
Area: Core
Comments Due: 10/31/2020
---

> # Native Command Error Handling

Reviewer 4: I'd like to see the PS team dogfood the implementation of this feature by using it instead of the Start-NativeExecution function used in the PS build module.

Reviewer: This was one of the motivations of doing this feature so we can remove that helper function

when executed in batch as opposed to interactive mode.

The `$PSUseNativeExitStatus` variable would govern whether exit status is used in determining the
boolean value of a native command for a conditional context. Possible values would be:

- `$false`: (the default) the boolean value of native command is defined as whether or not the
  length of the output stream is non-zero. This is the same as existing PowerShell treatment of this
  case.
- `$true`: the boolean value of a native command is it's exit status (`$?`), and

This would allow syntax like `if` and `while` to be usefully combined with native commands.

#### Add strict pipeline chain failure semantics.

Treat only ignored exit statuses as exceptions.

The `$PSStrictPipeLineChain` preference variable would govern the exit status in the last command of
a pipeline. Possible values would be:

- `$false`: (the default) would ignore `$false` exit status on the last command. This is the same as
  existing PowerShell treatment of this case.
- `$true`: for a pipeline that is being used in a conditional context (see above), and where the

>   exist status of the last command in the pipeline is `$false`, would create a terminating error

Reviewer 5: ```suggestion
  exit status of the last command in the pipeline is `$false`, would create a terminating error
```

---
RFC: RFC00XX
Author: Jason Helmick
Status: Draft
Area: Core
Comments Due: 10/31/2020
---

# Native Command Error Handling

PowerShell scripts using native commands would benefit from being able to use error handling
features like those used by cmdlets.

## Motivation

In PowerShell by default, script processing continues when non-terminating errors occur. This is a
benefit when expecting non-terminating errors in normal execution such as non-responsive computers from
a list. This default behavior is controlled with the preference variable
`$ErrorActionPreference` default of `Continue`.

During development and debugging, often customers prefer that script execution stops when a
non-terminating error occurs. PowerShell currently supports customers with this ability by setting
the preference variable in the script.

```Powershell
$ErrorActionPreference = 'Stop'
```

> Native commands return an exit code to the calling application which will be zero for success or

Reviewer 6: ```suggestion
Native commands usually return an exit code to the calling application which will be zero for success or
```

Reviewer 7: I would put the `usually` in `which will usually be zero for success`. I think every process must report an exit code

non-zero for failure. However, native commands currently do not participate in the PowerShell error
stream. Customers working with native commands in their scripts will need to check the execution
status after each call using a helper function similar to below:

```Powershell
if( ! $? )
{
    exit $lastexitcode;

> }

Reviewer 8: How `$LASTEXITCODE` would be stale if it's checked right after a native command call?

> }

Reviewer 7: Well the given snippet doesn't include a native command. If it did, then I imagine it's guaranteed to be fresh

 ```

Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`

> stale.

Reviewer 6: this might also argue for improving `$LASTEXITCODE` behavior so is less or not stale

To support a similar style of error handling with native commands, this specification proposes
converting native command errors into PowerShell errors similar to the `sh`-compatible

> `set -e` mode.

Reviewer 7: Consider something like:

> In POSIX shells, this need to terminate on command error is addressed by the [`set -e`](https://pubs.opengroup.org/onlinepubs/009695399/utilities/set.html) configuration, which causes the shell to exit when a command fails.
>
> This specification proposes a similar idea, but adapted to the PowerShell conventions of preference variables and catchable, self-describing, terminating error objects.

> `set -e` mode.

Reviewer 6: agreed, this behavior should probably be defined (as not all will have the correct understanding of `set -e`.

 The specification and alternative proposals are based on the
 [Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
 committee review of the associated
 [pull request](https://github.com/PowerShell/PowerShell/pull/3523), and
 [implementation plan](https://github.com/PowerShell/PowerShell-RFC/pull/88#issuecomment-613653678)

## Specification

> This RFC proposes including native commands in the error handling framework when the feature is

Reviewer 7: I think `error handling framework` is somewhat amorphous language that won't help us crystalise a clear specification here.

Instead, maybe something like:

> This RFC proposes a preference variable to configure the elevation of errors produced by native commands to first-class PowerShell errors, so that native command failures will produce error objects that are added to the error stream and may terminate execution of the script without added boilerplate.

> This RFC proposes including native commands in the error handling framework when the feature is

Reviewer 6: i don't have as much trouble with the initial language, but the proposed language is very clear and direct, perhaps both?

enabled by adding an error to the error stream if the exit code of a native command is non-zero.

The `$PSNativeCommandErrorAction` preference variable will implement a version of the

> `$ErrorActionPreference` variable for native commands.

Reviewer 7: Why does one start with `PS` and the other doesn't? That's, at best, an ugly and avoidable inconsistency. At worst I think it's confusing and counterintuitive.

- The value will default to `Ignore` for compatibility with existing behavior.
- For non-zero exit codes and except for the value 'Ignore', an `ErrorRecord` will be added to
  `$Error` that wraps the exit code and the command executed that returned the exit code.
- Initially, only the existing values of `$ErrorActionPreference` will be supported.
- The set of values may be extended later to include `MatchErrorActionPreference`, which should
  apply the `$ErrorActionPreference` setting to native commands.
- A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference`
  values, where `MatchErrorActionPreference` is converted to the current value of
  `$ErrorActionPreference`

> Valid values for `$ErrorActionPreference`

Reviewer 7: At this point I'm somewhat confused:

- We proposed a new preference variable and talked about one possible setting, which relays it to `$ErrorActionPreference`
- We also said the two variables should be convertible from one to the other, implying that `$PSNativeCommandErrorPreference` has one more member than `$ErrorActionPreference`
- But now we've stopped discussing the first preference variable and are describing `$ErrorActionPreference`, which already exists today with a known set of behaviours.

Author: Cleaned up section to address.

| Value           | Definition
----------------  | -------------------
| Break           | Enter the debugger when an error occurs or when an exception is raised.
| Continue        | (Default) - Displays the error message and continues executing.

> | Ignore          | Suppresses the error message and continues to execute the command. The Ignore value is intended for per-command use, not for use as saved preference. Ignore isn't a valid value for the $ErrorActionPreference variable.

Author: `Ignore` is corrected. will add statement regarding stderr

| Inquire         | Displays the error message and asks you whether you want to continue.
| SilentlyContinue| No effect. The error message isn't displayed and execution continues without interruption.
| Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream
| Suspend         | Automatically suspends a workflow job to allow for further investigation.

The reported error record should be created with the following details:

| Property        | Definition
----------------  | -------------------

> | exception:      | `ExitCode`, with the exit code of the failed command.

Reviewer: Seems like we should just create a new NativeCommandException type and have `ExitCode` be the property

> | error id:       | `"Program {0} failed with unhandled exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.

Reviewer: I think the message should be "Program {0} ended with non-zero exit code {1}"

| error category: | `ErrorCategory.NotSpecified`.

> | object:         | exit code

Reviewer: Perhaps have a `Source` property that is the full path to the app?

 This does not provide the actual semantics of bash `set -eo pipefail` as Bourne shell-style
 integration with existing status handling syntax is not implemented.

> `Set -eo pipefail` is a combination of `Set -u` and `Set -o pipefail`:

Reviewer 6: i think this is platform/shell specific - on my mac at least with bash, `set -eo pipefail` does not also `set -u` as the following transcript shows:
```powershell
PS> bash

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
bash-3.2$ set -eo pipefail
bash-3.2$ echo $SFDKLSF

bash-3.2$ set -u
bash-3.2$ echo $SFDKLSF
bash: SFDKLSF: unbound variable
PS>
```

Author: In light of the addition of set -o pipefail to the rfc, I cleaned up the definitions in the motivation section and removed set -eo.

- `Set -u`, a reference to any variable not previously defined results in an error. Similar to
  Set-StrictMode in PowerShell.
- `Set -o pipefail`, by default, pipeline success is determined by the last command executed. `Set
  -o pipefail` returns an error if any command in the pipeline fails.

 As a result of not including the semantics of `Set -eo pipefail`, PowerShell script logic for
 handling native command exit codes would need to use either `try`..`catch` or existing exit status

>  handling language constructs, according to the setting of this preference variable.

Reviewer 5: I think @SteveL-MSFT really wanted this included, as it was mentioned in some of the discussion of the original RFC and issue, but I think it can probably be stashed below in the Considerations.

>  handling language constructs, according to the setting of this preference variable.

Author: Removed and placed in the considerations section.

>  One way of overriding `$ErrorActionPreference` for a single native command and handling

Reviewer: You mea $PSNativeCommandErrorAction

>  it's exit status explicitly would be to put this logic into a script block and call it

Reviewer 8: ```suggestion
 its exit status explicitly would be to put this logic into a script block and call it
```

>  with the invocation operator (`&`).

Reviewer 7: This paragraph seems like a hanging addition, but could be turned into good scenario examples, like:

```powershell
try
{
    & { $PSNativeCommandErrorAction = 'Stop'; my_command $arg }
}
catch
{
    # Handle $_ here
}
```

Reviewer 2: @rjmholt, re:

> One thing that isn't discussed here is the actual error output of the native command.

Note the pending suggestion to allow `>` to target _variables_, which would enable collecting the stderr lines with `2> variable:errs`, as suggested by @JamesWTruher: https://github.com/PowerShell/PowerShell/issues/4332
What hasn't been fleshed out is the syntax  to _also_ pass the stderr _through_ (the way that `-ErrorVariable errs` implicitly does) and  (less urgently), how to _add_ to a preexisting target variable's value.

Reviewer 2: @rjmholt, re:

> `& { $PSNativeCommandErrorAction = 'Stop'; my_command $arg }`

Perhaps a simper idiom to recommend is the following:

```powershell
my_command $arg || $(throw)  #  Note the need for $(...)
```

Or, for scripts called from the outside:

```powershell
my_command $arg || $(exit $LASTEXITCODE)  #  Note the need for $(...)
```

Unfortunately, we can't be as concise as POSIX-like shells (`my_command $arg || exit`), both because we're saddled with `$(...)` and because `exit` requires `$LASTEXITCODE` in order to relay the exit code (which would otherwise default to `0`).

Conversely, it would be nice to have a concise way to _ignore_ failure _ad hoc_.
The following does that, but it is somewhat obscure:

```powershell
# Ignore non-zero exit code from my_command
my_command $arg || @()
```

POSIX-like shells offer the `:` built-in for this purpose, which is a quiet no-op that always succeeds.
We could consider offering it too.

## Alternative Approaches and Considerations

> We could extend this existing behavior to give equivalent functionality to bash `Set -eo pipefail`.

Reviewer 6: Are you suggesting that this work be undertaken? I'm having a bit of trouble untangling which is an alternative and what are considerations.

Author: WE agreed to add set -o pipefail to this rfc.

This includes:

- `Set -u`, a reference to any variable not previously defined results in an error. Similar to
  `Set-StrictMode` in PowerShell.
- `Set -o pipefail`, by default, pipeline success is determined by the last command executed. `Set
  -o pipefail` returns an error if any command in the pipeline fails.

> Today there exists a workaround for handling native command exit codes using a `try`..`catch` block.

Reviewer: It seems that the best practice for bash scripts is to enable `pipefail` so that if any native command in the pipeline fails, the script would exit.  An alternative is in the future we could support a `nopipefail` type option.

This alternative may be considered in the future.

### Add "strict" native command option

The `$PSStrictNativeCommand` preference should treat creation of an `ErrorRecord` for native
commands in the same way as this is treated elsewhere. Described here as a Boolean, could be
considered as an enum to allow for future expansion. Possible values are:

- `$false`: (the default) ignore non-zero exit codes. This is the same as existing PowerShell
  treatment of this case.
- `$true`: Populate the error stream of the native command with an `ErrorRecord` associated with an
  `ExitException` exception.

### Modifying existing semantics to consider exit code and exit status

> The error will throw an exception, potentially terminating the session, in the same situation as

Reviewer 8: ```suggestion
The error will throw an exception, potentially terminating the execution, in the same situation as
```

other non-terminating errors will do this, i.e. where `$ErrorActionPreference` is set to `"stop"`.
This would not be the desired behavior on commands where the script already handles a non-zero exit
code, which would require the addition of extra boilerplate to use multiple native commands in
combination. There are a number of ways that this could be made more flexible by integrating exit
code and exit status handling into the language syntax.

#### Convert non-terminating errors to terminating where the command output is used

This approach implements semantics equivalent to bash `set -eo pipefail` in the runtime layer.

> The `$PSStrictPipeLine` preference variable would govern promotion of a non-terminating error to a

Reviewer: Seems like we could just have a different value to the enum for `$PSNativeCommandErrorAction` instead of a new preference variable?

### Use lexical scope/Exception handling extensions

Dynamic scoping approach would have side-effects on called scripts. This is analogous to the
behavior of Bourne type shells, which maintain this behavior for
[historic compatibility](http://austingroupbugs.net/view.php?id=52) reasons.

Lexical scoping of native command error handling would improve on earlier options by integrating
native command error handling fully into the existing exception handling without any
side-effects in calling or called scripts.

With this approach, native command error handling mode would be used through language syntax instead
of preference variables or cmdlets. A possible syntax might be to add a strictness option to the try
statement which applies to the lexical scope of the try statement.

Lexical scope requires more runtime overhead; a mapping must be maintained at runtime between each
runtime scope and its corresponding lexical scope that existed at parse-time. In addition, lexical
scope may not apply well to errors. Errors produce stack traces at runtime and should be handled in
a runtime-facing way. For these reasons, this alternative is not under consideration.

> 

Reviewer 6: some of these would need a _bunch_ more explanation in order to pursue. I'm not sure what to do with this section. My assumption is that these aren't really part of the spec, just forestalling "what if" conversations?
